lib/LLVMIRCodeGen/libjit/libjit.cpp

```diff
@@ -728,53 +728,78 @@ static void libjit_arg_min_generic(const inpT *inpW, outT *outW,
 template <typename T>
 static void libjit_max_pool_generic(const T *inW, T *outW, const dim_t *inWdims,
                                     const dim_t *outWdims, dim_t *kernelSizes,
-                                    dim_t *strides, dim_t *pads) {
-  dim_t pad_t = pads[0];
-  dim_t pad_l = pads[1];
-  dim_t stride_h = strides[0];
-  dim_t stride_w = strides[1];
-  dim_t kernel_h = kernelSizes[0];
-  dim_t kernel_w = kernelSizes[1];
-  // For each sample in the batch:
-  for (dim_t n = 0; n < outWdims[0]; n++) {
-    // For each (x,y) step in the input/output tensor:
-    sdim_t x = -(sdim_t)pad_t;
-    for (dim_t ax = 0; ax < outWdims[1]; x += stride_h, ax++) {
-      sdim_t y = -(sdim_t)pad_l;
-      for (dim_t ay = 0; ay < outWdims[2]; y += stride_w, ay++) {
+                                    dim_t *strides, dim_t *pads, T defVal) {
 
-        // For each layer in the output tensor:
-        for (dim_t z = 0; z < inWdims[3]; z++) {
-          int first = 1;
-          T max = 0;
+  size_t kernelH = kernelSizes[0];
+  size_t kernelW = kernelSizes[1];
 
-          // For each element in the pool filter:
-          for (dim_t fx = 0; fx < kernel_h; fx++) {
-            for (dim_t fy = 0; fy < kernel_w; fy++) {
-              sdim_t ox = x + fx;
-              sdim_t oy = y + fy;
+  size_t strideH = strides[0];
+  size_t strideW = strides[1];
 
-              // Ignore index access below zero (this is due to padding).
-              if (ox < 0 || oy < 0 || ox >= (sdim_t)inWdims[1] ||
-                  oy >= (sdim_t)inWdims[2]) {
-                continue;
-              }
+  size_t padT = pads[0];
+  size_t padL = pads[1];
 
-              float val =
-                  inW[libjit_getXYZW(inWdims, n, (dim_t)ox, (dim_t)oy, z)];
+  // For each input in the batch.
+  for (size_t n = 0; n < inWdims[0]; n++) {
 
-              if (first || (val >= max)) {
-                first = 0;
-                max = val;
-              }
+    // For each output height.
+    ssize_t i_h_min = -(ssize_t)padT;
+    for (size_t o_h = 0; o_h < outWdims[1]; o_h++, i_h_min += strideH) {
+
+      // Effective kernel height limits.
+      ssize_t f_h_min = libjit_conv_flt_min(inWdims[1], kernelH, i_h_min);
+      ssize_t f_h_max = libjit_conv_flt_max(inWdims[1], kernelH, i_h_min);
+      ssize_t f_h_len = libjit_conv_flt_len(f_h_min, f_h_max);
```

Comment:
Added some description about the functions in `libjit_defs.h` where are implemented so that we know what is the meaning of the variables they compute.